src/k_smallest.rs

```diff
@@ -88,6 +88,45 @@ where
     storage
 }
 
+pub(crate) fn k_smallest_relaxed_general<I, F>(iter: I, k: usize, mut comparator: F) -> Vec<I::Item>
+where
+    I: Iterator,
+    F: FnMut(&I::Item, &I::Item) -> Ordering,
+{
+    if k == 0 {
+        iter.last();
+        return Vec::new();
+    }
+
+    let mut iter = iter.fuse();
+    let mut buf = iter.by_ref().take(2 * k).collect::<Vec<_>>();
+
+    if buf.len() < k {
+        buf.sort_unstable_by(comparator);
+        return buf;
+    }
+
+    buf.select_nth_unstable_by(k - 1, |a, b| comparator(a, b));
+    buf.truncate(k);
+
+    iter.for_each(|val| {
+        if comparator(&val, &buf[k - 1]) != Ordering::Less {
+            return;
+        }
+
+        buf.push(val);
```

Comment:
Before `push`, it might be valuable to add something like `assert!(buf.len() < 2 * k /*or buf.capacity()*/);` to ensure that the compiler cut part of `push` code (we know `buf` does not need to grow, and less code makes things faster as the compiler can optimize things).
It seems we don't have a benchmark for `k_smallest` but such benchmark would make it clear if it's valuable or not.

Reply:
In Tantivy - where the benchmarks from the linked article come from - this was indeed a measurable improvement, but the `assert` was insufficient and unsafe code was required to actually achieve the desired results, c.f.

https://github.com/quickwit-oss/tantivy/blob/99a59ad37e833e1129f7e1abd25738a0adfae73a/src/collector/top_score_collector.rs#L820-L830

Would you be open to such an implementation here?

Reply:
I went to source and `push` does `if self.len == self.buf.capacity() { reserve_for_push }` so maybe `assert_ne!(buf.len(), buf.capacity());` but `assert!(buf.len() < buf.capacity())` should really be enough.
About unsafe, you can read this https://github.com/rust-itertools/itertools/pull/560#issuecomment-2025715434 . So I would rather avoid unsafe code anywhere I can and I really think it should be possible here.

Reply:
It can be improved later if an `assert` really is not enough.

Reply:
I will give the assert a try when I add the benchmarks. But only later tonight when I hopefully have some spare time.

Reply:
The `assert_ne` variant measurably improves the benchmark results. I therefore included it in the second commit.